chapter05/mssw/exercise3.pl

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/perl
+use v5.10;
+use strict;
+use warnings;
+use Data::Dumper;
+
+my %total_bytes;
+
+my $filename = 'chapter05/mssw/coconet.dat';
+my $output   = 'chapter05/mssw/coconet-redump.dat';
+
+# my $filename = 'chapter05/mssw/coconet-sample.dat';  # smaller sample
+
+open( my $read_fh, '<:encoding(UTF-8)', $filename )
+  or die "could not open file '$filename' $!";
+
+while ( my $row = <$read_fh> ) {
+
+    # check if the line starts with a '#' and skip it
+    if ( $row =~ /^\s*#/ ) {
+        next;
+    }
+    my ( $source, $destination, $bytes ) = split /\s+/, $row;
+    $total_bytes{$source}{$destination} += $bytes;
+}
+close $read_fh;
+
+# Lets create a complex hash! \o/
+#  %report = (
+#   $source => {
+#     destinations = {
+#       $destination => $bytes,
+#       ...
+#     },
+#     total = $total_bytes,
+#   },
+#   ...
+# );
+my %report;
+foreach my $source ( sort keys %total_bytes ) {
+    for my $destination ( sort keys %{ $total_bytes{$source} } ) {
+        $report{$source}->{destinations}{$destination} +=
+          $total_bytes{$source}{$destination};
+        $report{$source}->{total} += $total_bytes{$source}{$destination};
+    }
+}
+
+open( my $write_fh, '>:encoding(UTF-8)', $output )
+  or die "could not create file '$output' $!";
+
+# Now lets redump the file
+foreach my $source ( sort { $report{$b} cmp $report{$a} } keys %report ) {
```

Comment:
This was supposed to sort based on the name of the source machine, but I think this is sorting it based on the hash ref values of the destination hash, not the key names. 
This should be: `sort { $a cmp $b} keys %report` or just `sort keys %report`